---
title: "2025_RNA_DESeq"
output: html_document
date: '2025-06-10'
editor_options: 
  chunk_output_type: console
---

## Note:
## Anything within the brackets "<>"" should be customised, fill in with relevant info/path, then remove the brackets and run the script 

```{r load packages}
#once packages are installed (only needs to be done once) you need to activate the libraries each session

packs_required<-c("BiocManager","pacman")
if_insta_packs<-packs_required %in% rownames(installed.packages())
if (any(if_insta_packs == FALSE)) {
  install.packages(packs_required[!if_insta_packs])
  }

# load CRAN and Bioconductor packages
pacman::p_load(tidyverse,rtracklayer,data.table,ggrepel,ggpubr,SummarizedExperiment,DESeq2,viridis,RColorBrewer,ggupset,ggtext,ggExtra,rstatix,patchwork,ggh4x,UpSetR)

```

```{r load files}
## Change this to your main project folder
PATH<-"<path/to/your/project/folder/>"
setwd(PATH) 


# load Raw counts into R, if your FeatureCounts Results file was saved in a different name, change it accordingly
counts <- read.delim2("<FeatureCounts_RNASeq>.txt", header = TRUE,skip = 1, row.names = 1 )

# load and format the annotation information
Annotation_file<-import("<ref_annotation>.gtf") %>% as.data.frame()
rownames(Annotation_file)<-Annotation_file$gene_id

# makesure the rownames(ie. gene_ids) match in between count table and annotation file, only proceed if the following command has an outcome of TURE
all.equal(rownames(counts), rownames(Annotation_file))

# Metadata, this need to be customised per dataset, where the first column matches names of the alignment files (i.e. the bam files), and additional column(s) for grouping information or treatment etc.

# In default, the count table will have 5 columns of meta data and the remaining being raw read counts that each sample has an individual column. So the starting column for counts is typically 6, if your count table is different, change the number here:
Starting_col<-6

### !!!IMPORTANT!!! If this is a new RNAseq run, skip (ie. comment out) the next 2 lines (add # in front of both lines) and load you own meta data instead
RNA_meta<-data.frame(Sample_ID=colnames(counts[Starting_col:length(colnames(counts))])) %>% 
  separate(Sample_ID, c("Strain","replicate","SeqID",NA), remove = FALSE)

### For new RNA seq analyses, change the RNA_meta template, and load it here (remove # in the following line to execute it). See the Data_analyses_on_Server_ReadMe file for instructions on how to construct the metadata file.
# RNA_meta<-read_csv("RNA_meta_template.csv",skip_empty_rows = TRUE)

# Set the control group, change the name here. Note this name needs to match one of the values from the Strain column in the RNA_meta file
Ctr_strain<-"<control_group_name>" 

# identify alternative strain
alt_strain<-unique(RNA_meta$Strain)
alt_strain<-alt_strain[alt_strain!=Ctr_strain]

# convert the Strain variable into factor and define levels, make sure the control group goes first
RNA_meta$Strain<-factor(RNA_meta$Strain, levels = c(Ctr_strain,alt_strain))


```


```{r Assemble all parts}

# Three parts are required to make a DESeq object, which would allow differential expression: count table, annotation info, and metadata

# assemble files
library(SummarizedExperiment)

se_test <- SummarizedExperiment(
  assays = list(counts = as.matrix(counts[,c(Starting_col:ncol(counts))])),
  rowRanges = as(Annotation_file, "GRanges"),
  colData = RNA_meta
)

```

```{r Differential expression}

# differential expression
library(DESeq2)
nrow(se_test)

# this will filter out genes has less than 5 reads, change this cut-off according to your results
se <- se_test[rowSums(assay(se_test, "counts")) > 5, ]
nrow(se)

# then rearrange the se object based on the reordered Strain levels
se <- se[, order(se$Strain)]
colData(se)

# saveRDS(se, "Summarised_Experiment_data.rds")


# if you have multiple grouping variables, can use multiple ones here design= ~group + treatment

# We need to adjust for the differences in library size between samples, to avoid drawing incorrect conclusions. The way this is typically done for RNA-seq data can be described as a two-step procedure. First, we estimate size factors - sample-specific correction factors such that if the raw counts were to be divided by these factors, the resulting values would be more comparable across samples. Next, these size factors are incorporated into the statistical analysis of the data. 

# switch our SummarizedExperiment object over to a DESeqDataSet as it has the internal structure to store these size factors. We also need to tell it our main experiment design, which is Strain

dds <- DESeq2::DESeqDataSet(se, design = ~ Strain) 

dds <- estimateSizeFactors(dds)

# DESeq2 and edgeR model RNA-seq counts as negative binomial distribution to account for a limited number of replicates per group, a mean-variance dependency and a skewed count distribution.
dds <- estimateDispersions(dds)

# We can use the nbinomWaldTest()function of DESeq2 to fit a generalized linear model (GLM) and compute log2 fold changes (synonymous with “GLM coefficients”, “beta coefficients” or “effect size”) corresponding to the variables of the design matrix. The design matrix is directly related to the design formula and automatically derived from it. 
dds <- nbinomWaldTest(dds)


############# normalized table to do pairwise comparison and export ############# 
# NOTE: DESeq2 doesn’t actually use normalized counts, rather it uses the raw counts and models the normalization inside the Generalized Linear Model (GLM). These normalized counts will be useful for downstream visualization of results, but cannot be used as input to DESeq2 or any other tools that peform differential expression analysis which use the negative binomial model.
# Some details regarding the DESeq methods of counts normalisation: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
Norm_rnaseq_table<-as.data.frame(counts(dds, normalized = TRUE))

# write.csv(Norm_rnaseq_table,"Norm_Counts_rnaseq.csv", row.names = TRUE)
# Note, once this csv file is saved, make a copy elsewhere, otherwise it would be replaced next time when you run the code.
```


```{r Explore the differential expression results}
# check the Deseq results names, if you have multiple treatment groups (eg. multiple DR strains), there should be results for each of them with the control group (eg. Wt)
resultsNames(dds)

# Look over the treatment group values DR strain to suit your results
DR_Wt_deseq<-NULL
# This will loop through all DR vs Wt results and combine them into one table
for (dr in alt_strain) {
  resTime <- results(dds, contrast = c("Strain", dr, Ctr_strain)) 
# summary(resTime)
  temp_df<-as.data.frame(resTime)
  temp_df$Strain_compare<-paste0(dr,"_",Ctr_strain)
  DR_Wt_deseq<-rbind(DR_Wt_deseq,temp_df)
}

# write.csv(DR_Wt_deseq,"<results_name_deseq>.csv", row.names = TRUE)
# Note, once this csv file is saved, make a copy elsewhere, otherwise it would be replaced next time when you run the code.

```
```